pkg/util/fswatch/fswatch.go

```diff
@@ -0,0 +1,146 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fswatch
+
+import (
+	"log"
+	"math"
+	"os"
+	"path/filepath"
+
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+)
+
+const (
+	// debounceDelay is the amount of time that must elapse after a file changes without any other file changes before
+	// a signal is sent.
+	debounceDelay = time.Second
+)
+
+// WatchDirRecursively recursively watches the given directory path. If any file under dir changes, a debounced signal
+// is sent on the returned channel. The debounce delay prevents flooding when multiple files are updated at once e.g.
+// during upgrade.
+func WatchDirRecursively(dir string) (<-chan struct{}, error) {
+	subdirs, err := recursiveSubdirs(dir)
+	if err != nil {
+		return nil, err
+	}
+
+	// Raw notifies that must be debounced.
+	notifyRaw := make(chan struct{})
+	for _, d := range subdirs {
+		if err := watchDir(d, notifyRaw); err != nil {
+			return nil, err
+		}
+	}
+
+	// Debounced notifies.
+	notify := make(chan struct{})
+	go func() {
+		timer := time.NewTimer(time.Duration(math.MaxInt64))
+		for {
+			select {
+			case <-notifyRaw:
+				if timer == nil {
+					timer = time.NewTimer(debounceDelay)
+					break
+				}
+				timer.Reset(debounceDelay)
+			case <-timer.C:
+				notify <- struct{}{}
+			}
+		}
+	}()
+
+	return notify, nil
+}
+
+func watchDir(dir string, notify chan<- struct{}) error {
+	var err error
+	doneInit := make(chan struct{}, 1)
+	go func() {
+		for {
+			// All errors are likely to be same reason here, just pick one.
+			var watcher *fsnotify.Watcher
+			watcher, err = fsnotify.NewWatcher()
+			if err != nil {
+				log.Print(err)
+				return
+			}
+			defer watcher.Close()
+
+			done := make(chan struct{}, 1)
+			go func() {
+				for {
+					select {
+					case event, ok := <-watcher.Events:
+						if !ok {
+							log.Printf("watcher channel closed for %s", dir)
+							done <- struct{}{}
+							return
+						}
+						if event.Op&fsnotify.Write == fsnotify.Write {
+							log.Println("modified file:", event.Name)
+							select {
+							case notify <- struct{}{}:
+							default:
```

Comment:
does this `select` block exist because sometimes multiple `WRITE` events can arrive for a single file change?
If yes, than a comment line explaining it would be great here because it's far from trivial.

Reply:
This is just avoiding blocking on sending the notify - added comment.